### PR TITLE
Refactored parser

### PR DESCRIPTION
**Pull request by disinvite**
Here it is! I went with a class-based approach this time, which (hopefully) helps code readability. Collecting global variables and vtables should work, but I haven't tested with an actual tool to verify the results. There are some new unit tests, including a lot of basic state machine tests to help drive the deisgn.

We now have syntax warnings (that the parser can ignore) and errors (that obstruct parsing). Both are reported in the `checkorder` tool. I think you had said that checkorder could evolve into a general-purpose decomp linter. We are probably at a good jumping off point for that now.

In the decomp code, I changed `// FUNCTION` to `// SYNTHETIC` in a few places in both ISLE and LEGO1 where appropriate. Right now, there are three types of markers that don't match:

- `// VTABLE: LEGO1 0x100dcb10 SYNTHETIC`
- `// GLOBAL: LEGO1 0x10102188 0x400` (only one ocurrence)
- `// DECOMP: ALPHA 0x1008b99d` (these two are from my PRs)

The last one is easy to fix/remove. For VTABLE SYNTHETICs, I think we need another indicator. (VTABLESYN?) If we did just take the following line (whatever it is) as the class name, the parser would get tripped up in this common case:

```
// VTABLE 0x1234
// SIZE 0x8
class Whatever : public Parent {
```

What do you think about the second case? In isledecomp/isle#308 I had talked about adding variable size or vtable length to the respective indicators. Is that something we want to support now? Should I have the parser just ignore them for the time being?

### CODE REVIEW

**Comment by foxtacles** [comment]
Great work! Just a few comments:

> `// VTABLE: LEGO1 0x100dcb10 SYNTHETIC`

Sorry, I forgot to remove `SYNTHETIC` from those annotations, that is a remnant of some automatic substitutions I ran.

I think the original idea was - correct me if I'm wrong @tahg - that the parser ought to support all cases irrespective if the line is followed by a comment or not, presumably by looking for the next `class ` or `struct ` token at the beginnings of the following lines (ignoring comments):

```
// VTABLE: LEGO1 0x100dc1b0
// SIZE stuff class struct <- line ignored
other stuff <- line ignored
// class MxCollection<MxVariable *> <- detects `class ` at the beginning of the line (comment/whitespace ignored)
```

```
// VTABLE: LEGO1 0x100dc1b0
// SIZE stuff class struct <- line ignored
other stuff <- line ignored
class MxCollection : Base { <- detects `class ` at the beginning of the line
```

I'm not sure how much effort that would be or how it would complicate things in general with regards to edge cases etc. Personally I think we should try to keep the parser code relatively straightforward and I wouldn't mind introducing a distinct primary marker in support of that, like you suggested, but I suppose if we can enable the above it would be more elegant.

We don't know yet if the cases ("commented class" or "implemented class") need to be reported differently depending on `reccmp` needs them ("comment" value or line number), but I guess having it as a boolean attribute to know the distinction just in case we need it doesn't hurt.

> In the decomp code, I changed // FUNCTION to // SYNTHETIC in a few places in both ISLE and LEGO1 where appropriate

Similarly, I believe the idea was that `FUNCTION` can *also* be followed by a comment with the function name, the same way `TEMPLATE` and `SYNTHETIC` work. (I think I got this wrong in my last comment in the original PR, sorry for that).

Since "synthetic" is meant to annotate functions that were not written by us but generated by the compiler (basically only the scalar ddtor for now), `SYNTHETIC` wouldn't be semantically appropriate for the inline functions that turned out not to be inlined.

In summary, I guess a generic approach that gives the same capability to all 4 function markers would be required, indicating in all cases if there is a comment in the next line - if yes, tools like `reccmp` take the comment value - if not, the "regular" case the way we use it most of the times happens, where the line number is used.

That means that all of the 4 function markers have a boolean attribute, something like `is_commented` (or the reverse, `is_implementation`) to know whether the name (comment value) or the line number is relevant for checks in `reccmp`.

> Is that something we want to support now? Should I have the parser just ignore them for the time being?

I think for now we should ignore it simply to keep the scope of the current efforts limited to what we have right now, and then address all the other concerns (of which we still have several) in separate PRs

File: LEGO1/mxpalette.cpp [hunk 1]
```diff
@@ -3,7 +3,8 @@
 #include "mxomni.h"
 #include "mxvideomanager.h"
 
-// GLOBAL: LEGO1 0x10102188 0x400
+// GLOBAL: LEGO1 0x10102188
+// SIZE: 0x400
```
Comment (foxtacles) [inline LEGO1/mxpalette.cpp]: Minor nitpick, but so far we didn't use a `:` for `SIZE`

File: tools/isledecomp/isledecomp/parser/util.py [hunk 2]
```diff
@@ -78,39 +51,45 @@ def is_blank_or_comment(line: str) -> bool:
...
-def distinct_by_module(offsets: List) -> List:
-    """Given a list of offset markers, return a list with distinct
-    module names. If module names (case-insensitive) are repeated,
-    choose the offset that appears first."""
+def is_marker_exact(line: str) -> bool:
+    return markerExactRegex.match(line) is not None
+
+
+template_class_decl_regex = re.compile(
+    r"\s*(?:\/\/)?\s*class (\w+)<([\w]+)\s*(\*+)?\s*>"
```
Comment (foxtacles) [inline tools/isledecomp/isledecomp/parser/util.py]: Should we consider `struct` as an alternative to `class`? I don't think there's any cases where we use structs with vtables but maybe in the future

Comment (disinvite) [reply tools/isledecomp/isledecomp/parser/util.py]: I looked and didn't find any places where we use it right now. No problem to add it, though

**Review by foxtacles — APPROVED** [review]
Looks good to me! Merging into parent PR, feel free to open another PR later if you want to make further changes. Thank you!

**Pull request merged**